wasmCloud/wasmcloud-k8s-operator
# Basic POST/DELETE implementation

Based upon the work that was done [here](https://github.com/wasmCloud/wasmcloud-k8s-operator/pull/2).

### Test
- To test this operator, create a local cluster using Kind or minikube 
- `make install && make run` this will register the CRD and start the operator on `localhost` talking to your local cluster
- Start a local nats server
- Run a fake lattice controller using `nats reply 'wasmbus.alc.default' '{"status":"received"}'`
- `kubectl apply -f config/samples/` and this is what the fake lattice controller should receive

```json
{
  "name": "wasmcloudapplication-sample",
  "namespace": "default",
  "application": {
    "kind": "WasmCloudApplication",
    "apiVersion": "core.wasmcloud.com/v1beta1",
    "metadata": {
      "name": "wasmcloudapplication-sample",
      "namespace": "default",
      "uid": "2b46172c-560c-4c28-9db1-3410cc4ff061",
      "resourceVersion": "4366620",
      "generation": 1,
      "creationTimestamp": "2021-09-14T16:05:49Z",
      "annotations": {
        "description": "This is my app",
        "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"core.wasmcloud.com/v1beta1\",\"kind\":\"WasmCloudApplication\",\"metadata\":{\"annotations\":{\"description\":\"This is my app\",\"version\":\"v0.0.1\"},\"name\":\"wasmcloudapplication-sample\",\"namespace\":\"default\"},\"spec\":{\"components\":[{\"name\":\"userinfo\",\"properties\":{\"image\":\"wasmcloud.azurecr.io/fake:1\"},\"traits\":[{\"properties\":{\"replicas\":4,\"spread\":[{\"name\":\"eastcoast\",\"requirements\":{\"zone\":\"us-east-1\"},\"weight\":80},{\"name\":\"westcoast\",\"requirements\":{\"zone\":\"us-west-1\"},\"weight\":20}]},\"type\":\"spreadscaler\"},{\"properties\":{\"target\":\"webcap\",\"values\":{\"port\":8080}},\"type\":\"linkdef\"}],\"type\":\"actor\"},{\"name\":\"webcap\",\"properties\":{\"contract\":\"wasmcloud:httpserver\",\"image\":\"wasmcloud.azurecr.io/httpserver:0.13.1\",\"link_name\":\"default\"},\"type\":\"capability\"},{\"name\":\"ledblinky\",\"properties\":{\"contract\":\"wasmcloud:blinkenlights\",\"image\":\"wasmcloud.azurecr.io/ledblinky:0.0.1\"},\"traits\":[{\"properties\":{\"replicas\":1,\"spread\":[{\"name\":\"haslights\",\"requirements\":{\"ledenabled\":true}}]},\"type\":\"spreadscaler\"}],\"type\":\"capability\"}]}}\n",
        "version": "v0.0.1"
      },
      "managedFields": [
        {
          "manager": "kubectl-client-side-apply",
          "operation": "Update",
          "apiVersion": "core.wasmcloud.com/v1beta1",
          "time": "2021-09-14T16:05:49Z",
          "fieldsType": "FieldsV1",
          "fieldsV1": {
            "f:metadata": {
              "f:annotations": {
                ".": {},
                "f:description": {},
                "f:kubectl.kubernetes.io/last-applied-configuration": {},
                "f:version": {}
              }
            },
            "f:spec": {
              ".": {},
              "f:components": {}
            }
          }
        },
        {
          "manager": "main",
          "operation": "Update",
          "apiVersion": "core.wasmcloud.com/v1beta1",
          "time": "2021-09-14T16:05:49Z",
          "fieldsType": "FieldsV1",
          "fieldsV1": {
            "f:status": {
              ".": {},
              "f:fromLatticeController": {}
            }
          }
        }
      ]
    },
    "spec": {
      "components": [
        {
          "name": "userinfo",
          "type": "actor",
          "properties": {
            "image": "wasmcloud.azurecr.io/fake:1"
          },
          "traits": [
            {
              "type": "spreadscaler",
              "properties": {
                "replicas": 4,
                "spread": [
                  {
                    "name": "eastcoast",
                    "requirements": {
                      "zone": "us-east-1"
                    },
                    "weight": 80
                  },
                  {
                    "name": "westcoast",
                    "requirements": {
                      "zone": "us-west-1"
                    },
                    "weight": 20
                  }
                ]
              }
            },
            {
              "type": "linkdef",
              "properties": {
                "target": "webcap",
                "values": {
                  "port": 8080
                }
              }
            }
          ]
        },
        {
          "name": "webcap",
          "type": "capability",
          "properties": {
            "contract": "wasmcloud:httpserver",
            "image": "wasmcloud.azurecr.io/httpserver:0.13.1",
            "link_name": "default"
          }
        },
        {
          "name": "ledblinky",
          "type": "capability",
          "properties": {
            "contract": "wasmcloud:blinkenlights",
            "image": "wasmcloud.azurecr.io/ledblinky:0.0.1"
          },
          "traits": [
            {
              "type": "spreadscaler",
              "properties": {
                "replicas": 1,
                "spread": [
                  {
                    "name": "haslights",
                    "requirements": {
                      "ledenabled": true
                    }
                  }
                ]
              }
            }
          ]
        }
      ]
    },
    "status": {
      "fromLatticeController": "received"
    }
  }
}
```
Note that this is the Application from https://github.com/wasmCloud/lattice-controller/tree/main/oam, but it is nested under the `application` key, and we have added `name` and `namespace` at the top level.

- `kubectl delete -f config/samples/` and this is what the fake lattice controller should receive

```json
{
  "name": "wasmcloudapplication-sample",
  "namespace": "default",
  "application": null
}
```

### Question
For deletion, we realised that we cannot just send an empty manifest to the controller so we wrapped everything. Does it make sense for us to send the `application` field as `null` while specifying the application name and namespace to the lattice controller, or can you recommend a better way to do this?


It is worth noting that we still have not implemented the OAM model and many other bits are missing. We are currently tracking this [here](https://github.com/wasmCloud/wasmcloud-k8s-operator/projects/1?fullscreen=true).

TODO before merging
- [ ] agree on protocol question, above
- [ ] address/backlog outstanding todos in the code
